raft/src/main/java/org/apache/kafka/raft/internals/UpdateVoterHandler.java

```diff
@@ -104,26 +111,39 @@ public CompletableFuture<UpdateRaftVoterResponseData> handleUpdateVoterRequest(
             );
         }
 
-        // KAFKA-16538 will implement the case when the kraft.version is 0
-        // Check that the cluster supports kraft.version >= 1
+        // Read the voter set from the log or leader state
         KRaftVersion kraftVersion = partitionState.lastKraftVersion();
-        if (!kraftVersion.isReconfigSupported()) {
-            return CompletableFuture.completedFuture(
-                RaftUtil.updateVoterResponse(
-                    Errors.UNSUPPORTED_VERSION,
-                    requestListenerName,
-                    new LeaderAndEpoch(
-                        localId,
-                        leaderState.epoch()
-                    ),
-                    leaderState.leaderEndpoints()
-                )
-            );
-        }
+        final Optional<KRaftVersionUpgrade.Voters> inMemoryVoters;
+        final Optional<VoterSet> voters;
+        if (kraftVersion.isReconfigSupported()) {
+            inMemoryVoters = Optional.empty();
 
-        // Check that there are no uncommitted VotersRecord
-        Optional<LogHistory.Entry<VoterSet>> votersEntry = partitionState.lastVoterSetEntry();
-        if (votersEntry.isEmpty() || votersEntry.get().offset() >= highWatermark.get()) {
+            // Check that there are no uncommitted VotersRecord
+            Optional<LogHistory.Entry<VoterSet>> votersEntry = partitionState.lastVoterSetEntry();
+            if (votersEntry.isEmpty() || votersEntry.get().offset() >= highWatermark.get()) {
```

Comment:
so if there is an uncommitted voters record this update is no-op (we wait until there are only committed voters records present before updating the voters?)

Reply:
Correct and the remote voter will retry because of the error type returned.